### PR TITLE
log_files: add blocking download_file; add erase log file function

### PR DESCRIPTION
**Pull request by dayjaby**
This PR adds two new functions to the log_files plugin:

- a blocking download_file
- an erase function

Note, that for the erase function, LOG_ERASE is being used. This mavlink message does not prescribe any feedback. To check whether the logs got really deleted, you can verify with get_entries.

Running the example looks like:

```
./MAVSDK/examples/logfile_download$ cmake -Bbuild -H. && cmake --build build -j4 && ./build/logfile_download udp://:14540
-- Configuring done
-- Generating done
-- Build files have been written to: ./MAVSDK/examples/logfile_download/build
[100%] Built target logfile_download
[01:09:06|Info ] MAVSDK version: v0.50.1 (mavsdk_impl.cpp:28)
Waiting to discover system...
[01:09:06|Info ] New system on: 127.0.0.1:58354 (with sysid: 1) (udp_connection.cpp:194)
[01:09:06|Debug] New: System ID: 1 Comp ID: 1 (mavsdk_impl.cpp:484)
[01:09:06|Debug] Component Autopilot (1) added. (system_impl.cpp:361)
[01:09:06|Debug] Component Unsupported component (240) added. (system_impl.cpp:361)
[01:09:07|Debug] Discovered 2 component(s) (system_impl.cpp:529)
[01:09:07|Debug] Received all entries (log_files_impl.cpp:169)
Got log file with ID 0 and date 2021-12-08T16:28:48Z
[01:09:07|Debug] 46080 B of 353765 B (320.16 kiB/s) (log_files_impl.cpp:418)
[01:09:07|Debug] 92160 B of 353765 B (237.55 kiB/s) (log_files_impl.cpp:418)
[01:09:07|Debug] 138240 B of 353765 B (333.791 kiB/s) (log_files_impl.cpp:418)
[01:09:08|Debug] 184320 B of 353765 B (328.368 kiB/s) (log_files_impl.cpp:418)
[01:09:08|Debug] 230400 B of 353765 B (326.467 kiB/s) (log_files_impl.cpp:418)
[01:09:08|Debug] 276480 B of 353765 B (319.142 kiB/s) (log_files_impl.cpp:418)
[01:09:08|Debug] 322560 B of 353765 B (317.143 kiB/s) (log_files_impl.cpp:418)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Error] Ignoring wrong offset (log_files_impl.cpp:359)
[01:09:08|Debug] 353765 B of 353765 B (306.913 kiB/s) (log_files_impl.cpp:418)
```
```

### CODE REVIEW

File: src/plugins/log_files/include/plugins/log_files/log_files.h [hunk 1]
```diff
@@ -153,12 +153,28 @@ class LogFiles : public PluginBase {
...
+     * @return Result of the log download.
+     */
+
+    LogFiles::Result download_log_file(LogFiles::Entry entry, std::string path);
+
+    /**
+     * @brief Erase all log files.
      */
-    void download_log_file_async(Entry entry, std::string path, DownloadLogFileCallback callback);
+    void erase_log_files();
```
Comment (julianoes) [inline src/plugins/log_files/include/plugins/log_files/log_files.h:177]: Note that the plugin.h and plugin.cpp files are auto-generated from proto.
If you did in fact auto-generate it from proto, then you just need to make a PR with the proto additions as well here: https://github.com/mavlink/MAVSDK-Proto and then update the submodule in this repo.

See: https://mavsdk.mavlink.io/main/en/cpp/contributing/plugins.html#add-api-to-proto

File: src/plugins/log_files/log_files_impl.cpp [hunk 2]
```diff
@@ -203,6 +203,21 @@ void LogFilesImpl::list_timeout()
...
+        std::make_shared<std::promise<LogFiles::Result>>();
+    auto future_result = prom->get_future();
+
+    download_log_file_async(entry, file_path, [prom](LogFiles::Result result, LogFiles::ProgressData progress) {
+        UNUSED(progress);
+        if (result != LogFiles::Result::Next) {
+            prom->set_value(result);
+	}
+    });
+    return future_result.get();
```
Comment (julianoes) [inline src/plugins/log_files/log_files_impl.cpp:218]: This is fine, thanks for that. To enable it in the API you can remove this line here: https://github.com/mavlink/MAVSDK-Proto/blob/11769a4ee6f3f9b0fef71532f773d40ab3e7d55c/protos/log_files/log_files.proto#L17

**Review by julianoes — COMMENTED** [review]
Thanks for the PR @dayjaby!

**Comment by dayjaby** [comment]
@julianoes thanks for the review! One more question: should this erase log files be a "fire and forget"-function, like in native MAVLink...or does it make sense to call get_entries after a short amount of time and check whether they got deleted or not? And provide some feedback with either `LogFiles::Success` or `LogFiles::Failure`?

**Comment by julianoes** [comment]
Yer, that's the question. I'm not sure, mostly because the list function takes quite a bit of time. You could add a comment that says "if you want to be sure the log has been deleted, you have to list it again and check".

File: src/plugins/log_files/include/plugins/log_files/log_files.h (hunk 1 above)
Comment (dayjaby) [reply src/plugins/log_files/include/plugins/log_files/log_files.h]: In fact I did not. Thanks for the links; I've added a PR to the protos repo.

**Comment by julianoes** [comment]
@dayjaby I think this would still be worthwhile. @dayjaby would you mind fixing the conflicts and adding the comment about the fact that that you need to check the logfiles again after deleting to make sure it worked.

**Comment by julianoes** [comment]
@dayjaby I've rebased this, fixed it up, and force pushed it. Let's hope it's now mergeable.

File: examples/logfile_download/logfile_download.cpp [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+            }
+        }
+        if (!download_failure && remove_log_files) {
+            /*
+             * If you want to be sure the log has been deleted, call get_entries again
+             * that there are no log files present anymore
+             *
+             * TODO: provide a more reliable solution
+             */
+            log_files.erase_log_files();
```
Comment (JonasVautherin) [inline examples/logfile_download/logfile_download.cpp:114]: I am a bit surprised that this compiles. Shouldn't it be `erase_all_log_files`? As per the proto file?

Comment (dayjaby) [reply examples/logfile_download/logfile_download.cpp:114]: You seem to be right. I use `drone.log_files.erase_all_log_files` in python as well. Don't we catch compilation errors in the examples automatically?

Comment (julianoes) [reply examples/logfile_download/logfile_download.cpp:114]: I've fixed this all but probably didn't check the examples. Let me check.

**Review by JonasVautherin — APPROVED** [review]

**Comment by julianoes** [comment]
CI failure probably unrelated.

**Pull request merged**